vignettes/multiple-factors.Rmd

```diff
@@ -0,0 +1,196 @@
+colnames(fs_dat) <- c("fs_ind60", "fs_dem60")
+# alambda %*% attr(fs_dat, "acov")[[1]] %*% t(alambda)
+my2spa <- '
+   # latent variables (indicated by factor scores)
+     ind60 =~ 0.95538579 * fs_ind60 + 0.05816994 * fs_dem60
+     dem60 =~ 0.01834111 * fs_ind60 + 0.86888887 * fs_dem60
+   # constrain the errors
+     fs_ind60 ~~ 0.034595518 * fs_ind60
+     fs_dem60 ~~ 0.090775708 * fs_dem60
+     fs_ind60 ~~ 0.004017388 * fs_dem60
+   # latent variances
+     ind60 ~~ ind60
+     dem60 ~~ dem60
+   # regressions
+     dem60 ~ ind60
+'
+tspa_fit <- sem(model = my2spa, data  = fs_dat)
+standardizedSolution(tspa_fit)
+```
+
+which is more consistent with the SEM results.
+
+## Three-factor model example
+
+```{r}
+# CFA
+cfa_3fac <-  '
+  # latent variables
+  ind60 =~ x1 + x2 + x3
+  dem60 =~ y1 + y2 + y3 + y4
+  dem65 =~ y5 + y6 + y7 + y8
+'
+cfa_3fac_fit <- cfa(model = cfa_3fac,
+                    data  = PoliticalDemocracy,
+                    std.lv = TRUE)
+# A matrix
+pars <- lavInspect(cfa_3fac_fit, what = "est")
+lambda_mat <- pars$lambda
+psi_mat <- pars$psi
+sigma_mat <- cfa_3fac_fit@implied$cov[[1]]
+ginvsigma <- MASS::ginv(sigma_mat)
+alambda <- psi_mat %*% crossprod(lambda_mat, ginvsigma %*% lambda_mat)
+alambda
+```
+
+We can also use `R2spa::get_fs()`:
+
+```{r}
+(fs_dat_3fac <- get_fs(PoliticalDemocracy, model = cfa_3fac, std.lv = TRUE)) |> 
+  head()
+```
+
+Therefore, we need to specify cross-loadings when using 2S-PA.
+
+```{r}
+tspa_fit_3fac <- tspa(model = "dem60 ~ ind60
+              dem65 ~ ind60 + dem60",
+              data = fs_dat_3fac,
+              vc = attr(fs_dat_3fac, "av_efs"),
+              cross_loadings = attr(fs_dat_3fac, "fsA"))
+cat(attr(tspa_fit_3fac, "tspaModel"))
+parameterestimates(tspa_fit_3fac)
+```
+
+
+```{r include=FALSE, eval=FALSE}
+# Compute factor scores using lavPredict
+fs_3fac_dat <- lavPredict(cfa_3fac_fit, acov = "standard")
+colnames(fs_3fac_dat) <- c("fs_ind60", "fs_dem60", "fs_dem65")
+# alambda %*% attr(fs_dat, "acov")[[1]] %*% t(alambda)
+mod_2spa_3fac <- '
+   # latent variables (indicated by factor scores)
+     ind60 =~ 0.95064774 * fs_ind60 + (-0.02069724) * fs_dem60 + 
+              0.09868528 * fs_dem65
+     dem60 =~ (-0.005967124) * fs_ind60 + 0.533020047 * fs_dem60 +
+              0.401095944 * fs_dem65
+     dem65 =~ 0.02951139 * fs_ind60 + 0.41603787 * fs_dem60 +
+              0.49133377 * fs_dem65
+   # constrain the errors
+     fs_ind60 ~~ 0.0340104952 * fs_ind60 + 0.0003881641 * fs_dem60 + 
+                 0.005026024 * fs_dem65
+     fs_dem60 ~~ 0.0586870314 * fs_dem60 + 0.0533752458 * fs_dem65
+     fs_dem65 ~~ 0.051573299 * fs_dem65
+   # latent variances
+     ind60 ~~ ind60
+     dem60 ~~ dem60
+     dem65 ~~ dem65
+   # regressions
+     dem60 ~ ind60
+     dem65 ~ ind60 + dem60
+'
+tspa_fit_3fac <- sem(model = mod_2spa_3fac, data = fs_3fac_dat)
+standardizedSolution(tspa_fit_3fac)
+```
+
+Compare to SEM:
+
+```{r}
+sem_3fac <- sem("
```

Comment:
to match the structure of the passing model, should there be a tab in front of each line?

Reply:
This is more a stylistic difference, but ya let's do that.